src/hooks/useNotificationManager.tsx

```diff
@@ -0,0 +1,165 @@
+    privatePostsUserIds: [],
+    setNotificationsRead: () => {},
+    lastReadAt: undefined,
+  });
+
+export const NotificationsManagerProvider = ({
+  children,
+}: {
+  children?: React.ReactNode;
+}) => {
+  const [lastNotificationReadTime, setLastNotificationReadTime] =
+    useAsyncStorage('lastNotificationReadTime');
+  const { data, refetch } = useNotifications();
+
+  /**
+   * We need a ref here so we can have access to the current value inside the
+   * closure created grabbing the last read time from async storage
+   */
+  const lastReadAtRef = useRef<NotificationsManagerContextType['lastReadAt']>();
+  const [lastReadAt, setLastReadAt] =
+    useState<NotificationsManagerContextType['lastReadAt']>();
+
+  const updateLastReadAt = useCallback((readAt: Date | null) => {
+    setLastReadAt(readAt);
+    lastReadAtRef.current = readAt;
+  }, []);
+
+  /**
+   * Load persisted lastReadAt from async storage
+   */
+  useEffect(() => {
+    if (lastNotificationReadTime.isFetchedAfterMount) {
+      updateLastReadAt(
+        lastNotificationReadTime.data
+          ? new Date(lastNotificationReadTime.data)
+          : null,
+      );
+    }
+  }, [
+    lastNotificationReadTime.data,
+    lastNotificationReadTime.isFetchedAfterMount,
+    updateLastReadAt,
+  ]);
+
+  /**
+   * Persist lastReadAt to async storage on background/close
+   */
+  useEffect(() => {
+    const handleAppStateChange = (newState: AppStateStatus) => {
+      if (newState === 'background' || newState === 'inactive') {
+        lastReadAtRef.current &&
+          setLastNotificationReadTime(lastReadAtRef.current.toISOString());
+      }
+      if (newState === 'active') {
+        refetch();
+      }
+    };
+
+    const handleOnBlur = () => {
+      lastReadAtRef.current &&
+        setLastNotificationReadTime(lastReadAtRef.current.toISOString());
+    };
+
+    const listener = AppState.addEventListener('change', handleAppStateChange);
+
+    // Android can be forced killed without triggering an AppState change
+    // Luckily there is a blur event emitted
+    if (Platform.OS === 'android') {
+      const blurListener = AppState.addEventListener('blur', handleOnBlur);
+      return () => {
+        listener.remove();
+        blurListener.remove();
+      };
+    }
+
+    return () => {
+      listener.remove();
+    };
+  }, [refetch, setLastNotificationReadTime]);
+
+  useEffect(() => {
+    onNotificationReceived(() => {
+      refetch();
+    });
+  }, [refetch]);
+
+  const notifications =
+    data?.notificationsForUser.edges.map((edge) => edge.node) ?? [];
+
+  const setNotificationsRead = useCallback(() => {
+    updateLastReadAt(new Date());
+  }, [updateLastReadAt]);
+
+  let unreadCount: number | undefined;
+  let privatePostsUserIds: string[] | undefined;
+
+  const notificationsToAuthorId = (_notifications: FeedNotification[]) =>
+    compact(
+      _notifications.map((n) => {
+        if (n.__typename === 'PrivatePostNotification') {
```

Comment:
If we make `__typename` a property of each notification type, then we won't need to do casting if we first check the type name.

Reply:
Done